2020-04-23-Rust-1.43.0.md (#56)
rust-lang-ru/translations: ru/announcements/2020-04-23-Rust-1.43.0.md

# Выпуск Rust 1.43.0

Команда Rust рада сообщить о выпуске новой версии, 1.43.0. Rust — это язык программирования, позволяющий каждому создавать надёжное и эффективное программное обеспечение.

Если вы установили предыдущую версию Rust средствами `rustup`, то для обновления до версии 1.43.0 вам достаточно выполнить следующую команду:

```console
rustup update stable
```

Если у вас ещё не установлен `rustup`, вы можете [установить его](https://www.rust-lang.org/install.html) с соответствующей страницы нашего веб-сайта, а также посмотреть [подробные примечания к выпуску](https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1430-2020-04-23) на GitHub.

## Что вошло в стабильную версию 1.43.0

> В данном релиз достаточно мало значительных изменений, он не содержит нововведений. Мы стабилизировали некоторые новые API, улучшили производительность компилятора и добавили некоторые изменения в макросы. Чтобы узнать об изменениях, не вошедших в анонс, смотрите [подробные примечания к выпуску](https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1430-2020-04-23).

Reviewer: > В данном релиз ...

В данном релиз**е** ...

Reviewer 2: done

> В данном релиз достаточно мало значительных изменений, он не содержит нововведений. Мы стабилизировали некоторые новые API, улучшили производительность компилятора и добавили некоторые изменения в макросы. Чтобы узнать об изменениях, не вошедших в анонс, смотрите [подробные примечания к выпуску](https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1430-2020-04-23).

Reviewer: > В данном релиз достаточно мало значительных изменений, он не содержит нововведений.

В данном релизе достаточно мало значительных изменений (и или запятая) без нововведений.

### Фрагменты `item`

> Вы можете использовать фрагменты `item` встраивать `item` в тело типажей, реализаций или внешних блоков. Например:

Reviewer: Вы можете использовать фрагменты `item` встраивая `item` в тело типажей, реализаций или внешних блоков.

Reviewer: Также предлагаю повторное написание `item` убрать.

Reviewer 3: Вы можете использовать фрагменты `item` встраивая их в тело типажей, ...

Reviewer 2: done

```rust
macro_rules! mac_trait {
    ($i:item) => {
        trait T { $i }
    }
}
mac_trait! {
    fn foo() {}
}
```

### Type inference around primitives

The type inference around primitives, references, and binary operations was improved. A code sample makes this easier to understand: this code fails to compile on Rust 1.42, but compiles in Rust 1.43.

```rust
let n: f32 = 0.0 + &0.0;
```

In Rust 1.42, you would get an error that would say "hey, I don't know how to add an `f64` and an `&f64` with a result of `f32`." The algorithm now correctly decides that both `0.0` and `&0.0`s should be `f32`s instead.

### Изменения в библиотеке

> [Мы можем напрямую использовать ассоциированные к целым числами и числам с плавающей точкой константы](https://github.com/rust-lang/rust/pull/68952/), вместо того, чтобы импортировать модуль. Теперь мы можем писать `u32::MAX` или `f32::NAN` без `use std::u32;` и `use std::f32;`.

Reviewer: > Мы можем напрямую использовать **ассоциированные к целым числами и числам с плавающей точкой константы**

Каждый раз читая это "спотыкаюсь". Сложно :(

> # Выпуск Rust 1.43.0: улучшение производительности компилятора и вывода типов для примитивов, item в макросах, 

Reviewer 2: done

Команда Rust рада сообщить о выпуске новой версии, 1.43.0. Rust — это язык программирования, позволяющий каждому создавать надёжное и эффективное программное обеспечение.

Если вы установили предыдущую версию Rust средствами `rustup`, то для обновления до версии 1.43.0 вам достаточно выполнить следующую команду:

```console
rustup update stable
```

Если у вас ещё не установлен `rustup`, вы можете [установить его] с соответствующей страницы нашего веб-сайта, а также посмотреть [подробные примечания к выпуску] на GitHub.

## Что вошло в стабильную версию 1.43.0

> В данном релизе достаточно мало значительных изменений, он не содержит нововведений. Мы стабилизировали несколько новых API, улучшили производительность компилятора и добавили некоторые изменения в макросы. Чтобы узнать об изменениях, не вошедших в анонс, смотрите [подробные примечания к выпуску](https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1430-2020-04-23).

Reviewer 4: ```suggestion
Значительных изменений и нововведений этот релиз не содержит. Мы стабилизировали несколько новых API, улучшили производительность компилятора и добавили некоторые изменения в систему макросов. Для более подробной информации смотрите [подробные примечания к выпуску](https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1430-2020-04-23).
```

> В данном релиз достаточно мало значительных изменений, он не содержит нововведений. Мы стабилизировали некоторые новые API, улучшили производительность компилятора и добавили некоторые изменения в макросы. Чтобы узнать об изменениях, не вошедших в анонс, смотрите [подробные примечания к выпуску](https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1430-2020-04-23).

Reviewer 2: We have some new stabilized APIs

> В данном релиз достаточно мало значительных изменений, он не содержит нововведений. Мы стабилизировали некоторые новые API, улучшили производительность компилятора и добавили некоторые изменения в макросы. Чтобы узнать об изменениях, не вошедших в анонс, смотрите [подробные примечания к выпуску](https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1430-2020-04-23).

Reviewer 2: несколько новых [методов] API

> В данном релиз достаточно мало значительных изменений, он не содержит нововведений. Мы стабилизировали некоторые новые API, улучшили производительность компилятора и добавили некоторые изменения в макросы. Чтобы узнать об изменениях, не вошедших в анонс, смотрите [подробные примечания к выпуску](https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1430-2020-04-23).

Reviewer 5: api все таки интерфейс, мб просто использовать "методы" вместо него?

### Фрагменты `item`

> Вы можете использовать фрагменты `item` встраивая их в тело типажей, реализаций или внешних блоков. Например:

Reviewer 4: ```suggestion
Теперь вы можете использовать фрагменты `item`, встраивая их в тело типажей, реализаций или внешних блоков. Например:
```

```rust
macro_rules! mac_trait {
    ($i:item) => {
        trait T { $i }
    }
}
mac_trait! {
    fn foo() {}
}
```

Этот вызов сгенерирует следующий код:

```rust
trait T {
    fn foo() {}
}
```

### Вывод типа для примитивов

> Был улучшен вывод типа для примитивов, ссылок и бинарных операций. Пример кода облегчает понимание: этот код не компилируется в Rust 1.42, но компилируется в Rust 1.43.

Reviewer 3: Был улучшен вывод типа для примитивов, ссылок и бинарных операций. Например, следующий код не компилируется в Rust 1.42, но компилируется в Rust 1.43.

Reviewer 2: done

> Был улучшен вывод типа для примитивов, ссылок и бинарных операций. Пример кода облегчает понимание: этот код не компилируется в Rust 1.42, но компилируется в Rust 1.43.

Reviewer 2: done

> Был улучшен вывод типа для примитивов, ссылок и бинарных операций. Пример кода облегчает понимание: этот код не компилируется в Rust 1.42, но компилируется в Rust 1.43.

Reviewer 2: done

```rust
let n: f32 = 0.0 + &0.0;
```

> В Rust 1.42 вы получили бы сообщение об ошибке "Эй, я не знаю, как сложить `f64` и `&f64` с результатом типа `f32`" (casting `&f64` as `f32` is invalid - cannot cast `&f64` as `f32`). Теперь алгоритм правильно выводит, что оба значения `0.0` и `&0.0` должны быть типа `f32`.

Reviewer 4: ```suggestion
Rust 1.42 выдаёт ошибку "casting `&f64` as `f32` is invalid - cannot cast `&f64` as `f32`". Начиная с этого выпуска алгоритм правильно выводит тип операндов - `f32`.
```

> В Rust 1.42 вы получили бы сообщение об ошибке "Эй, я не знаю, как сложить `f64` и `&f64` с результатом типа `f32`" (casting `&f64` as `f32` is invalid - cannot cast `&f64` as `f32`). Теперь алгоритм правильно решает, что оба `0.0` и `&0.0` должны быть типа `f32`.

Reviewer 3: Теперь алгоритм правильно решает, что оба *значения* `0.0` и `&0.0`

Reviewer 2: done

> В Rust 1.42 вы получили бы сообщение об ошибке "Эй, я не знаю, как сложить `f64` и `&f64` с результатом типа `f32`" (casting `&f64` as `f32` is invalid - cannot cast `&f64` as `f32`). Теперь алгоритм правильно решает, что оба `0.0` и `&0.0` должны быть типа `f32`.

Reviewer 2: на сколько увидел в gitlocalize, его вручную добавили

> В Rust 1.42 вы получили бы сообщение об ошибке "Эй, я не знаю, как сложить `f64` и `&f64` с результатом типа `f32`" (casting `&f64` as `f32` is invalid - cannot cast `&f64` as `f32`). Теперь алгоритм правильно решает, что оба `0.0` и `&0.0` должны быть типа `f32`.

Reviewer 2: в оригинале нет сообщения об ошибке

> В Rust 1.42 вы получили бы сообщение об ошибке "Эй, я не знаю, как сложить `f64` и `&f64` с результатом типа `f32`" (casting `&f64` as `f32` is invalid - cannot cast `&f64` as `f32`). Теперь алгоритм правильно решает, что оба `0.0` и `&0.0` должны быть типа `f32`.

Reviewer 2: done

### Новая переменная окружения Cargo для тестов

> Чтобы помочь в проведении интеграционного тестирования, [Cargo теперь устанавливает новые переменные окружения](https://github.com/rust-lang/cargo/pull/7697).

Reviewer 3: Может *Cargo теперь устанавливает новую переменную окружения*? Все-таки одну переменную добавили

Reviewer 2: done

> Это легче всего объяснить на примере: допустим мы работаем на консольным проектом, назовём его "cli". Если мы пишем интеграционные тесты, то мы хотим запустить `cli` и посмотреть на результат её работы. Во время выполнения тестов и бенчмарков, Cargo устанавливает переменную окружения с именем `CARGO_BIN_EXE_cli`, и мы можем использовать её в нашем тесте:

Reviewer 4: ```suggestion
Это легче всего объяснить на примере: допустим, мы работаем над консольным проектом, назовём его "cli". В интеграционных тестах мы хотим запустить `cli` и посмотреть на результат её работы. Во время выполнения тестов и бенчмарков Cargo устанавливает переменную окружения с именем `CARGO_BIN_EXE_cli`, и мы можем использовать её в нашем тесте:
```

> Это легче всего объяснить на примере: допустим мы работаем на консольным проектом, назовём его "cli". Если мы пишем интеграционные тесты, то мы хотим вызвать `cli` и посмотреть, что она сделает. Во время выполнения тестов и бенчмарков, Cargo устанавливает переменную окружения, зовущуюся `CARGO_BIN_EXE_cli`, и мы можем использовать её внутри нашего теста:

Reviewer 3: ... Cargo устанавливает переменную окружения под именем `CARGO_BIN_EXE_cli`, и мы можем ...

Reviewer 2: сделал "переменную окружения с именем"

> Это легче всего объяснить на примере: допустим мы работаем на консольным проектом, назовём его "cli". Если мы пишем интеграционные тесты, то мы хотим вызвать `cli` и посмотреть, что она сделает. Во время выполнения тестов и бенчмарков, Cargo устанавливает переменную окружения, зовущуюся `CARGO_BIN_EXE_cli`, и мы можем использовать её внутри нашего теста:

Reviewer 2: сделал "мы хотим запустить cli", так как про бинарь говорят

> Это легче всего объяснить на примере: допустим мы работаем на консольным проектом, назовём его "cli". Если мы пишем интеграционные тесты, то мы хотим вызвать `cli` и посмотреть, что она сделает. Во время выполнения тестов и бенчмарков, Cargo устанавливает переменную окружения, зовущуюся `CARGO_BIN_EXE_cli`, и мы можем использовать её внутри нашего теста:

Reviewer 2: -> "посмотреть на результат её работы"

> Это легче всего объяснить на примере: допустим мы работаем на консольным проектом, назовём его "cli". Если мы пишем интеграционные тесты, то мы хотим вызвать `cli` и посмотреть, что она сделает. Во время выполнения тестов и бенчмарков, Cargo устанавливает переменную окружения, зовущуюся `CARGO_BIN_EXE_cli`, и мы можем использовать её внутри нашего теста:

Reviewer 2: done

> Это легче всего объяснить на примере: допустим мы работаем на консольным проектом, назовём его "cli". Если мы пишем интеграционные тесты, то мы хотим вызвать `cli` и посмотреть, что она сделает. Во время выполнения тестов и бенчмарков, Cargo устанавливает переменную окружения, зовущуюся `CARGO_BIN_EXE_cli`, и мы можем использовать её внутри нашего теста:

Reviewer 2: done

```rust
let exe = env!("CARGO_BIN_EXE_cli");
```

> Это упрощает вызов `cli`, так как теперь у нас есть прямой путь до него.

Reviewer 4: ```suggestion
`cli` становится легче вызвать, ведь теперь у нас в распоряжении полный его полный путь.
```

### Изменения в библиотеке

> [Мы можем напрямую использовать ассоциированные к целым числами и числам с плавающей точкой константы], вместо того, чтобы импортировать модуль. Теперь мы можем писать `u32::MAX` или `f32::NAN` без `use std::u32;` и `use std::f32;`.

Reviewer 3: Мы можем напрямую использовать ассоциированные к целым ~~числами~~ числам

Reviewer 2: сделал "Мы можем напрямую использовать константы, ассоциированные с целыми числами и числами с плавающей точкой"

> [Мы можем напрямую использовать ассоциированные к целым числами и числам с плавающей точкой константы], вместо того, чтобы импортировать модуль. Теперь мы можем писать `u32::MAX` или `f32::NAN` без `use std::u32;` и `use std::f32;`.

Reviewer 2: проверил окончания, сделал "Мы можем напрямую использовать константы, ассоциированные с целыми числами и числами с плавающей точкой"

> Также добавлен [новый модуль `primitive`](https://github.com/rust-lang/rust/pull/67637/) который реэкспортирует примитивные типы Rust. Он может пригодиться, когда вы пишите макрос и хотите быть уверенным, что типы не будут затенены.

Reviewer 4: ```suggestion
Также добавлен [новый модуль `primitive`](https://github.com/rust-lang/rust/pull/67637/), реэкспортирущий примитивные типы Rust. Он может пригодиться в написании макросов, чтобы быть уверенным, что типы не будут затенены.
```

> Ещё мы стабилизировали шесть новых API:

Reviewer 2: (имеется в виду этот тред: https://github.com/rust-lang-ru/translations/pull/56/files/112106d700489a3df7d4abfb93f76094c6346783#r414713977)

- [`Once::is_completed`]
- [`f32::LOG10_2`]
- [`f32::LOG2_10`]
- [`f64::LOG10_2`]
- [`f64::LOG2_10`]
- [`iter::once_with`]

### Другие изменения

> [Синтаксис](https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1430-2020-04-23), [пакетный менеджер Cargo] и [анализатор Clippy] также претерпели некоторые изменения.

Reviewer 5: Яб заменил анализатор на линтер, пусть англицизм, зато конкретное понятие

Reviewer 2: это уже по шаблону из почти всех предыдущих публикацией копируется, как и введение и спасибо участникам
